presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkNativeTaskRdd.java

```diff
@@ -0,0 +1,175 @@
+ */
+package com.facebook.presto.spark.classloader_interface;
+
+import org.apache.spark.MapOutputTracker;
+import org.apache.spark.Partition;
+import org.apache.spark.ShuffleDependency;
+import org.apache.spark.SparkContext;
+import org.apache.spark.SparkEnv;
+import org.apache.spark.TaskContext;
+import org.apache.spark.rdd.RDD;
+import org.apache.spark.rdd.ShuffledRDD;
+import org.apache.spark.rdd.ShuffledRDDPartition;
+import org.apache.spark.rdd.ZippedPartitionsPartition;
+import org.apache.spark.shuffle.ShuffleHandle;
+import org.apache.spark.storage.BlockId;
+import org.apache.spark.storage.BlockManagerId;
+import scala.Tuple2;
+import scala.collection.Iterator;
+import scala.collection.Seq;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.spark.classloader_interface.ScalaUtils.emptyScalaIterator;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+import static scala.collection.JavaConversions.asJavaCollection;
+import static scala.collection.JavaConversions.seqAsJavaList;
+
+/**
+ * PrestoSparkTaskRdd represents execution of Presto stage, it contains:
+ * - A list of shuffleInputRdds, each of the corresponding to a child stage.
+ * - An optional taskSourceRdd, which represents ALL table scan inputs in this stage.
+ * <p>
+ * Table scan is present when joining a bucketed table with an unbucketed table, for example:
+ * Join
+ * /  \
+ * Scan  Remote Source
+ * <p>
+ * In this case, bucket to Spark partition mapping has to be consistent with the Spark shuffle partition.
+ * <p>
+ * When the stage partitioning is SINGLE_DISTRIBUTION and the shuffleInputRdds is empty,
+ * the taskSourceRdd is expected to be present and contain exactly one empty partition.
+ * <p>
+ * The broadcast inputs are encapsulated in taskProcessor.
+ */
+public class PrestoSparkNativeTaskRdd<T extends PrestoSparkTaskOutput>
+        extends PrestoSparkTaskRdd<T>
+{
+    public static <T extends PrestoSparkTaskOutput> PrestoSparkNativeTaskRdd<T> create(
+            SparkContext context,
+            Optional<PrestoSparkTaskSourceRdd> taskSourceRdd,
+            // fragmentId -> RDD
+            Map<String, RDD<Tuple2<MutablePartitionId, PrestoSparkMutableRow>>> shuffleInputRddMap,
+            PrestoSparkTaskProcessor<T> taskProcessor)
+    {
+        requireNonNull(context, "context is null");
+        requireNonNull(taskSourceRdd, "taskSourceRdd is null");
+        requireNonNull(shuffleInputRddMap, "shuffleInputRdds is null");
+        requireNonNull(taskProcessor, "taskProcessor is null");
+        List<String> shuffleInputFragmentIds = new ArrayList<>();
+        List<RDD<Tuple2<MutablePartitionId, PrestoSparkMutableRow>>> shuffleInputRdds = new ArrayList<>();
```

Comment:
Use ImmutableList.builder and when requiring a list, call builder.build() to get an unmodifiable list.